lib/AST/Decl.cpp

```diff
@@ -3885,10 +3885,29 @@ findExistentialSelfReferences(CanGenericSignature existentialSig, Type type,
     return info;
   }
 
-  // Opaque result types of protocol extension members contain an invariant
-  // reference to 'Self'.
-  if (type->is<OpaqueTypeArchetypeType>())
-    return SelfReferenceInfo::forSelfRef(TypePosition::Invariant);
+  // If the signature of an opaque result type has a same-type constraint
+  // that refereces Self, it's invariant.
+  if (auto opaque = type->getAs<OpaqueTypeArchetypeType>()) {
```

Comment:
Nice catch! I think we can ignore opaque types altogether: it looks like the code you added to `typeEraseCovariant` would also erase stuff like `some Collection<Self>` to `any Collection`.

Reply:
Did this in https://github.com/apple/swift/pull/41183/commits/98ecb9b2029568ca9d7c7db68af34c027a932891

Reply:
It doesn't, because there's a same-type constraint between the opaque result type's `Element` type and `Self` that involves `Self`, so we reject attempts to open the existential when calling a function that returns `some Collection<Self>`.

Reply:
> so we reject attempts to open the existential when calling a function that returns `some Collection<Self>`

Type-erasing the resulting opaque archetype to `any Collection` sounds safe as far as I'm concerned. Am I missing the point, or is it something we can do in a follow-up PR?

Reply:
My concern is that we would erase to `any Collection` today, but then when we have existentials like `any Collection<String>`, we'll break code by changing the result of type inference here.